### ISSUE TITLE
Router is deployed with SNAT enabled regardless of enable_snat = "false" 

### ISSUE THREAD

**Issue by jgeorgeson** [terraform-providers/terraform-provider-openstack#280]
### Terraform Version

```bash
$ terraform -v
Terraform v0.11.5
+ provider.openstack v1.3.0

Your version of Terraform is out of date! The latest version
is 0.11.7. You can update by downloading from www.terraform.io/downloads.html
```

### Affected Resource(s)

openstack_networking_router_v2

### Terraform Configuration Files

```hcl
#
# Define provider info 
#
provider "openstack" {
  auth_url    = "${var.openstack_auth_url}"
  insecure    = "${var.openstack_insecure}"
  domain_name = "${var.project_domain_name}"
  tenant_name = "${var.project_name}"
}

#
# Lookup various resource IDs
#

data "openstack_networking_network_v2" "transit_network" {
  name = "${var.openstack_transit_network}"
}

data "openstack_networking_subnetpool_v2" "pool23" {
  name = "${var.openstack_subnet_pool}"
}

#
# Create a network in our project
# project is defined in provider.tf
#
resource "openstack_networking_network_v2" "tenant_network" {
  region         = "${var.openstack_region}"
  name           = "tenant-net-${var.project_name}"
  shared         = "false"
  admin_state_up = "true"
}

#
# Create a subnet in our new network(s)
resource "openstack_networking_subnet_v2" "tenant_subnet" {
  name            = "tenant-subnet-${var.project_name}-01"
  tenant_id       = "${var.project_id}"
  region          = "${var.openstack_region}"
  network_id      = "${openstack_networking_network_v2.tenant_network.id}"
  cidr            = "${var.subnet_cidr}"
  subnetpool_id   = "${data.openstack_networking_subnetpool_v2.pool23.id}"
  dns_nameservers = "${var.name_server}"

  ip_version = 4
}

resource "openstack_networking_router_v2" "tenant_router" {
  name                = "default-${var.project_name}"
  tenant_id           = "${var.project_id}"
  admin_state_up      = "true"
  distributed         = "true"
  enable_snat         = "false"
  external_network_id = "${data.openstack_networking_network_v2.transit_network.id}"
}

#
# Attach the Router to our Network via an Interface
#
resource "openstack_networking_router_interface_v2" "router_interface" {
  region    = "${var.openstack_region}"
  router_id = "${openstack_networking_router_v2.tenant_router.id}"
  subnet_id = "${openstack_networking_subnet_v2.tenant_subnet.id}"

  depends_on = [
    "openstack_networking_subnet_v2.tenant_subnet",
  ]
}
```

variables.tf:

```hcl
#
# Define the varialbles within your Terraform configuration 
# update/replace variables to match your deployment  
# Provider - OpenStack
#

variable "openstack_auth_url" {
  description = "The endpoint url to connect to OpenStack."
  default     = "https://openstack.example.com:5000/v3"
}

variable "openstack_insecure" {
  type        = "string"
  description = "Accept self signed SSL certificates?"
  default     = "true"
}

variable "openstack_region" {
  description = "Target Region"
  default     = "nova"
}

variable "openstack_az" {
  description = "Target Availability Zone"
  default     = "nova"
}

variable "openstack_transit_network" {
  description = "Openstack external Network Name"
  default     = "VIOINT_Transit_Network"
}

variable "openstack_subnet_pool" {
  description = "Name of the subnet pool"
  default     = "VIOINT_Tenant_pool23"
}

variable "project_domain_name" {
  description = "The name of the Domain."
  default     = "default"
}

variable "project_name" {
  description = "The name of the Project."
}

variable "project_id" {
  description = "The ID of the Project"
}

variable "subnet_cidr" {
  description = "CIDR for new subnet"
}

variable "name_server" {
  description = "name servers to be used for VMs"
  default     = ["...", "..."]
}
```

### Expected Behavior
The router should have SNAT disabled.

### Actual Behavior
The router had SNAT enabled.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

### Important Factoids
We are running VMWare Integrated OpenStack, and OpenStack release is Ocata. If I run `terraform apply` a second time with no other actions/changes, it notices that the deployed resource is out of sync with the request config and correctly disables SNAT.

**Comment by fatmcgav** [comment]
@jgeorgeson Are you able to Gist a Terraform debug log covering the creation of the router? As TF will read back the resource after the creation, so would be interesting to see what's being returned from your Openstack installation... 

`TF_LOG=debug OS_DEBUG=1 terraform apply...`

Be sure to strip out any sensitive info!

**Comment by jgeorgeson** [comment]
Sure, https://gist.github.com/jgeorgeson/c46299bf04d6a0e2df63f6d42aa5d895

Also this was generated with Terraform v0.11.7 so have now tested with latest too. :)

**Comment by jgeorgeson** [comment]
Oh I forgot the OS_DEBUG variable, will redo it.

**Comment by jgeorgeson** [comment]
Have rerun with `OS_DEBUG=1`. I also ran it with `-target openstack_networking_router_v2.tenant_router` to reduce the output. And here's a gist of the second apply, which disables SNAT.

https://gist.github.com/jgeorgeson/8d34d655040acecd5b9049b061d70dbc

**Comment by jtopjian** [comment]
Yep, looks like this is a bug. Because of how the `enable_snat` argument is being read in, it will only ever be sent in the request if it's set to `true`. Setting it to `false` causes the area of code to be ignored.

However, it looks like OpenStack routers have SNAT enabled by default. In my openstack environment, setting it to `false` gives me a error that disabling snat is not allowed by policy. Just so you aren't waiting for this bug to be fixed and release, you can make sure you're allowed to disable snat by doing the following:

```shell
$ neutron router-create foo
$ neutron router-gateway-set --disable-snat foo <public network>
```

In either case, we'll get this fixed shortly :)

**Comment by jgeorgeson** [comment]
We are indeed using that `router-gateway-set` command on every new router. 

> Setting it to false causes the area of code to be ignored

In that case it seems doubly odd that SNAT is correctly disabled by simply running `terraform apply` a second time.

**Comment by jtopjian** [comment]
> In that case it seems doubly odd that SNAT is correctly disabled by simply running terraform apply a second time.

That area of code checks for a change in value and any change (whether true to false or false to true) is sent in the code, so it's just the Create area that needs updated. :)

**Comment by jtopjian** [comment]
@jgeorgeson Quick update on this: I just merged a large update to this provider (#287) which will provide support to properly fix this bug. Because the vendor update was so large, I want to give some time to make sure it doesn't cause any bugs. So this will probably stay open for another 1-2 weeks. In the meantime, though, the above workaround should continue working for you.

I apologize for the delay.
